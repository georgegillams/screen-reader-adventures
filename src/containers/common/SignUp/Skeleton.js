import React from 'react';
import PropTypes from 'prop-types';
import { SectionSkeleton, InputSkeleton, ButtonSkeleton } from 'gg-components/Skeletons';

const Skeleton = props => {
  const { className } = props;

  const outerClassNames = [];

  if (className) {
    outerClassNames.push(className);
  }

  return (
    <div className={outerClassNames.join(' ')}>
      <SectionSkeleton />
      <InputSkeleton style={{ marginTop: '1.2rem' }} />
      <ButtonSkeleton />
    </div>
  );
};

Skeleton.propTypes = {
  className: PropTypes.string,
};

Skeleton.defaultProps = {
  className: null,
};

export default Skeleton;
