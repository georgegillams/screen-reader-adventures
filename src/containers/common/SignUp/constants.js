export const KEY = 'sign-up';
export const SIGN_UP = `app/${KEY}/SIGN_UP`;
