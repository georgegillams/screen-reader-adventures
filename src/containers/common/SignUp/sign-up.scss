@import '~utils/tokens';

.sign-up {
  &__link {
    margin-top: $spacing-base;
  }

  &__form {
    margin-top: $spacing-base;
  }
}
