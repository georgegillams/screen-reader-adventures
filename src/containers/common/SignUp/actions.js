import { createRoutine } from 'redux-saga-routines';

import { SIGN_UP } from './constants';

export const signUp = createRoutine(SIGN_UP);
