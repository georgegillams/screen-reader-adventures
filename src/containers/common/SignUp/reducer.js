import produce from 'immer';

import { signUp } from './actions';

export const initialState = {
  signingUp: false,
  signUpError: null,
  signUpResult: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case signUp.TRIGGER:
        draft.credentials = payload;
        draft.signUpResult = null;
        break;

      case signUp.REQUEST:
        draft.signingUp = true;
        draft.signUpError = null;
        break;

      case signUp.SUCCESS:
        draft.signingUp = false;
        draft.signUpResult = payload;
        break;

      case signUp.FAILURE:
        draft.signingUp = false;
        draft.signUpError = payload;
        break;
    }
  });

export default reducer;
