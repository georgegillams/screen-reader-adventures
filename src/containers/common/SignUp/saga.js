import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';
import { selectState } from './selectors';

import { setUser } from 'containers/common/Authenticator/actions';
import { signUp } from './actions';

export function* doSignUp() {
  const currentState = yield select(selectState());
  const { credentials } = currentState;

  const requestURL = apiStructure.signUp.fullPath;

  try {
    yield put(signUp.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(credentials),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(signUp.failure(result));
    } else {
      yield put(signUp.success(result));
      yield put(setUser.trigger(result));
    }
  } catch (err) {
    yield put(signUp.failure(err));
  } finally {
    yield put(signUp.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(signUp.TRIGGER, doSignUp);
}
