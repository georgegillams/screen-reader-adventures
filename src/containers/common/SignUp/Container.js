import React, { useState } from 'react';
import PropTypes from 'prop-types';
import TextLink from 'components/common/TextLink';
import PageTitle from 'components/common/PageTitle';
import { DebugObject } from 'components/common/DebugObject';
import { LoadingCover } from 'gg-components/LoadingCover';

import Skeleton from './Skeleton';

import { LoggedOutOnly } from 'components/common/Walls';
import { SignUpForm } from 'components/common/Forms';
import CookiesRequired from 'containers/common/CookiesRequired';
import { CONSENT_STATE_ALLOWED } from 'containers/common/Consent/constants';
import STYLES from './sign-up.scss';
import { cssModules } from 'gg-components/helpers/cssModules';

const getClassName = cssModules(STYLES);

const SignUp = props => {
  const [credentials, setCredentials] = useState({});

  const {
    signUp,

    consentState,
    signUpState,
    authenticatorState,

    className,
  } = props;
  const outerClassNames = [];

  if (className) {
    outerClassNames.push(className);
  }
  let preSubmitText = null;
  if (signUpState.signUpResult && signUpState.signUpResult.success) {
    preSubmitText = signUpState.signUpResult.success;
  }
  if (signUpState.signUpError && signUpState.signUpError.errorMessage) {
    preSubmitText = signUpState.signUpError.errorMessage;
  }

  // We shouldn't allow a user to sign up unless cookies are consented to
  const page = (
    <div className={outerClassNames.join(' ')}>
      <CookiesRequired reason={'sign up'} />
      <LoggedOutOnly user={authenticatorState.user}>
        <PageTitle name="Sign up">
          <SignUpForm
            className={getClassName('sign-up__form')}
            disabled={consentState.cookieConsent !== CONSENT_STATE_ALLOWED}
            loading={signUpState.signingUp}
            credentials={credentials}
            onDataChanged={setCredentials}
            onSubmit={() => {
              if (consentState.cookieConsent === CONSENT_STATE_ALLOWED) {
                signUp(credentials);
              }
            }}
            preSubmitText={preSubmitText}
          />
          <TextLink className={getClassName('sign-up__link')} href="/login">
            Already got an account? Login here.
          </TextLink>
        </PageTitle>
      </LoggedOutOnly>
    </div>
  );

  return (
    <>
      <LoadingCover loadingSkeleton={Skeleton} loading={authenticatorState.user === undefined} error={false}>
        {page}
      </LoadingCover>
      <DebugObject
        debugTitle="SignUp"
        debugObject={{
          signUp,
          authenticatorState,
          signUpState,
        }}
      />
    </>
  );
};

SignUp.propTypes = {
  signUp: PropTypes.func.isRequired,
  signUpState: PropTypes.shape({
    signingUp: PropTypes.bool,
    signUpError: PropTypes.object,
    signUpResult: PropTypes.object,
  }).isRequired,
  authenticatorState: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
  consentState: PropTypes.shape({
    cookieConsent: PropTypes.string,
  }).isRequired,
  className: PropTypes.string,
};

SignUp.defaultProps = {
  authenticatorState: null,
  signUpState: null,
  className: null,
};

export default SignUp;
