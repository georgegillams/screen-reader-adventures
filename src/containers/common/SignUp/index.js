import { memo } from 'react';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { signUp } from './actions';
import { selectState } from './selectors';
import { selectState as selectAuthenticatorState } from 'containers/common/Authenticator/selectors';
import { selectState as selectConsentState } from '../Consent/selectors';
import SignUp from './Container';
import injectSaga from 'client-utils/common/redux/inject-saga';
import injectReducer from 'client-utils/common/redux/inject-reducer';

import { KEY } from './constants';
import saga from './saga';
import reducer from './reducer';

const mapStateToProps = createStructuredSelector({
  signUpState: selectState(),
  consentState: selectConsentState(),
  authenticatorState: selectAuthenticatorState(),
});

export function mapDispatchToProps(dispatch) {
  return {
    signUp: payload => dispatch(signUp(payload)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);
const withSaga = injectSaga({ key: KEY, saga });
const withReducer = injectReducer({ key: KEY, reducer });

export default compose(withSaga, withReducer, withConnect, memo)(SignUp);
