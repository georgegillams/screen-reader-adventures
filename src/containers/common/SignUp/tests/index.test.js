import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState } from '../reducer';
import { initialState as initialAuthenticatorState } from 'containers/common/Authenticator/reducer';
import { initialState as initialConsentState } from '../../Consent/reducer';

import configureStore from 'client-utils/common/redux/configure-store';

import SignUpIndex from '../index';
import SignUp from '../Container';
import { CONSENT_STATE_ALLOWED } from 'containers/common/Consent/constants';

describe('<SignUp />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUpIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with initial state', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
          consentState={{ ...initialConsentState }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with signUpError', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
            signUpError: { error: 'not_found', errorMessage: 'No matching email found' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
          consentState={{ ...initialConsentState, cookieConsent: CONSENT_STATE_ALLOWED }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with signUpResult', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
            signUpResult: { uname: 'uname', email: 'email' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
          consentState={{ ...initialConsentState, cookieConsent: CONSENT_STATE_ALLOWED }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with signingUp=true', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
            signingUp: true,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
          consentState={{ ...initialConsentState, cookieConsent: CONSENT_STATE_ALLOWED }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with user', () => {
    const { container } = render(
      <Provider store={store}>
        <SignUp
          signUp={spy}
          signUpState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
          consentState={{ ...initialConsentState, cookieConsent: CONSENT_STATE_ALLOWED }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
