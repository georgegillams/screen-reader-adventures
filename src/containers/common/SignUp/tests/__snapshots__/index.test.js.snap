// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> should render correctly - index 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="skeleton__outer skeleton__section"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__input"
      style="margin-top: 1.2rem;"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with initial state 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="skeleton__outer skeleton__section"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__input"
      style="margin-top: 1.2rem;"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with signUpError 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="skeleton__outer skeleton__section"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__input"
      style="margin-top: 1.2rem;"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with signUpResult 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="skeleton__outer skeleton__section"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__input"
      style="margin-top: 1.2rem;"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with signingUp=true 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="skeleton__outer skeleton__section"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__input"
      style="margin-top: 1.2rem;"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<SignUp /> should render correctly with user 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="section__outer"
    >
      <h2
        class="section__heading"
        id="you're-logged-in"
      >
        You're logged in
      </h2>
      <span
        class="paragraph"
      >
        <a
          class="text-link text-link--text-link text-link--no-padding text-link--inline"
          href="/account"
        >
          Go to your account.
        </a>
      </span>
    </div>
  </div>
</div>
`;
